doubtfire-lms/doubtfire-api
# Feature/jplag implementation

Depends on [ doubtfire-lms/doubtfire-deploy#28 ](https://github.com/doubtfire-lms/doubtfire-deploy/pull/28)
Depends on [ doubtfire-lms/doubtfire-web#873 ](https://github.com/doubtfire-lms/doubtfire-web/pull/873)
Feasibility document: [ thoth-tech/documentation#528 ](https://github.com/thoth-tech/documentation/pull/528)

# Description

This branch adds JPLAG software plagiarism detection to OnTrack. The `unit_similarity_module.rb` file has been modified to extend the `check_similarity function` to perform a docker exec command to generate a jplag report for each task with "done" files in a unit.
It extracts the code files from the "done" zip files and temporarily stores them in `tmp/jplag/unitcode/task` while the jplag report is being generated.

## Type of change


# How Has This Been Tested?

- Use the modified docker-compose and added jplag.Dockerfile from [ Feature/add jplag plagiarism detection #28 ](https://github.com/doubtfire-lms/doubtfire-deploy/pull/28)
- Create a new task in a unit and have at least 2 student users upload similar code files for the task
- Change task due date to be in the past, so submissions can be accepted in the check_similarity check
- Run `rake submission:generate_pdfs` while cd'ed in doubtfire-api. This will create the "done" files.
- Run `rake submission:check_plagiarism` while cd'ed in doubtfire-api. This will run jplag on all units and tasks.

# Checklist:
